### DISCUSSION TITLE
`Unable to generate pydantic-core schema for <class 'datetime.datetime'>` when using Freezegun

### DISCUSSION THREAD

**Discussion by theadam** [pydantic/pydantic#9343] [Question] [answered]
In our unit tests, we sometimes use freezegun to make `datetime.now` deterministic to test out some specific functionality in some pydantic serializable models.  Whenever using freezegun I get this error:

```
pydantic.errors.PydanticSchemaGenerationError: Unable to generate pydantic-core schema for <class 'datetime.datetime'>. Set `arbitrary_types_allowed=True` in the model_config to ignore this error or implement `__get_pydantic_core_schema__` on your type to fully support it.
```

I tried patching `__get_pydantic_core_schema__` on the FakeDatetime and FakeDate classes in freezegun, but it doesnt seem to help.  Has anyone encountered this and found a solution?

**Comment by zachaysan** [comment]
I've also been hit with this issue. What was the change that caused this error to crop up?

**Reply by theadam** [reply]
@zachaysan If you are asking about the change we made, we are trying to transition to using pydantic for some model serialization.  We are currently using custom serialization.  Introducing pydantic is what caused this.

**Reply by zachaysan** [reply, marked as answer]
Thanks for letting me know @theadam. We were able to work around this issue which you can see here:

https://github.com/Flagsmith/flagsmith/pull/3876

It basically loads the model up before FreezeGun gets involved.

**Reply by theadam** [reply]
@zachaysan I think this solves it.  We are using dataclasses and TypeAdapters, but if I create the TypeAdapter before freezing time things are working! Thank you!

**Comment by zachaysan** [comment]
Happy to hear that it worked out for you :)

**Comment by lmignon** [comment]
I'm having the same problem using fastapi's testclient.... 
The only ugly working solution I've found is to patch the `match_type` method of the `GenerateSchema` class.

```python

from pydantic._internal._generate_schema import GenerateSchema
from pydantic_core import core_schema

initial_match_type = GenerateSchema.match_type


def match_type(self, obj):
    if getattr(obj, "__name__", None) == "datetime":
        return core_schema.datetime_schema()
    return initial_match_type(self, obj)


GenerateSchema.match_type = match_type
```

**Reply by admarko** [reply]
@lmignon -- where did you put this solution? in your pydantic class? In the test file? I'm having this same issue with fastapi/pytest/pydantic but can't figure out how to use your solution

**Reply by guibeira** [reply]
I put it on my conftest, and works great.

**Comment by disrupted** [comment]
I just started running into this issue after updating Pydantic from 2.8.2 to 2.9.2. Downgrading seems to resolve it.

**Comment by MishaVyb** [comment]
For me this solution works fine. I took implementation for `FutureDate` / `PastDate` and remove any `now_op=...` restrictions.

```python
if TYPE_CHECKING:
    DateType = Annotated[date, ...]
else:

    class DateType:

        @classmethod
        def __get_pydantic_core_schema__(
            cls, source: type[Any], handler: GetCoreSchemaHandler
        ) -> core_schema.CoreSchema:
            if cls is source:
                # used directly as a type
                return core_schema.date_schema()
            else:
                schema = handler(source)
                annotated_type = schema["type"]
                if annotated_type != "date":
                    raise PydanticUserError(
                        f"'{cls.__name__}' cannot annotate '{annotated_type}'.",
                        code="invalid-annotated-type",
                    )

                return schema

        def __repr__(self) -> str:
            return "Date"
```

**Reply by MishaVyb** [reply]
I think similar solutions will work for datetime

**Comment by avilaton** [comment]
We migrated our tests to use `time-machine`, a drop-in replacement for freezegun which doesn't mess with imports. It also turns out to be a bit more reliable and faster. We have now removed the hacks we had to do for freezegun.

https://github.com/adamchainz/time-machine

**Reply by benjsec** [reply]
Sadly time-machine has decided to start patching the monotoic clock without an option to disable, so breaks a lot of async tests. We're looking at moving from time-machine to freezegun to get around that problem and found this one instead

**Reply by Giuzzilla** [reply]
> Sadly time-machine has decided to start patching the monotoic clock without an option to disable, so breaks a lot of async tests. We're looking at moving from time-machine to freezegun to get around that problem and found this one instead

Just found out about this thread, just FYI (or for anyone looking) apparently `time-machine` fixed/reverted that behaviour in the 3.0.0 release (2025-11-18) so async should now work correctly

**Reply by zmievsa-semasoftware** [reply]
Pydantic logfire breaks for me when using time-machine

**Reply by beda42** [reply]
Just switched to time-machine without problems. Thanks for the suggestion.
